Various exec fixes

ncdc opened the pull request:
If stdin is noninteractive, the io.Copy from stdin to remoteStdin will
unblock when it finishes reading from stdin. In this case, make sure to
close remoteStdin so the server knows the client won't be sending any
more data. This ensures that the remote process terminates. For example:

```
echo foo | kubectl exec -i <pod> -- cat
```

Without this change, the `cat` process never terminates and `kubectl
exec` hangs.

Fix interactive exec sessions hanging after you type 'exit'.

Add e2e test to cover noninteractive stdin: `echo a | kubectl exec -i <pod>
cat`

Add e2e test to cover psuedo-interactive stdin: `kubectl exec -i <pod> bash`

Prep for sending multiple data frames over multiple streams in remote command
test, which is more likely to find flakes (requires bump of spdystream
once an issue with the frame worker queues not being fully drained when
a goaway frame is received).

Fixes #13394
Fixes #13395

**Release note info**
Fixed hanging when using non-interactive stdin data, e.g.  `echo foo | kubectl exec -i <pod> -- cat`
Fixed interactive `kubectl exec` sessions hanging after you type 'exit'

ncdc commented:
@smarterclayton @pmorie @liggitt

liggitt commented on test/e2e/kubectl.go:
```diff
@@ -159,11 +159,19 @@ var _ = Describe("Kubectl client", func() {
+			By("executing a command in the container with noninteractive stdin")
+			execOutput = newKubectlCommand("exec", fmt.Sprintf("--namespace=%v", ns), "-i", simplePodName, "cat").
+				withStdinData("abcd1234").
+				exec()
```
+1 for tests for this scenario

liggitt replied:
actually, it would be easy to add a `withStdin(io.Reader)` helper so we could also test an interactive exec with a  stdin that sends "echo hello\nexit\n", then blocks

ncdc replied:
Good idea. I'll definitely want to add that. Once I figure out why interactive is broken...

ncdc replied:
Blocks via `select {}` or something else?

liggitt replied:
Yeah, something like http://play.golang.org/p/zasWBNstx6

Commit:
bump(fsouza/go-dockerclient): 76fd6c6

liggitt commented on pkg/client/unversioned/remotecommand/remotecommand.go:
```diff
@@ -194,12 +194,19 @@ func (e *Streamer) doStream() error {
 		if err != nil {
 			return err
 		}
-		defer remoteStdin.Reset()
```
don't you still need the deferred remoteStdin.Reset() in the case where the remoteStdin `cp` never completes, or does the conn.Close() cover that?

ncdc replied:
I'll check and see, but it's broken until #13324 lands. I can test w/that locally.

liggitt replied:
actually, I'm surprised the remote connection was being held open with the deferred Reset() in place previously

ncdc replied:
It's a bit of a chicken and egg game. With this scenario:

    echo foo | kubectl exec -i <pod> cat

The `cat` command won't terminate until it gets EOF from its stdin. Its stdin comes from the spdy stream. The spdy stream gets its stdin from the non interactive input (`echo foo`). While `foo` is fully read and closed, `doStream` won't exit until `cat` is done, but we aren't resetting `remoteStdin` until `doStream` exits, etc etc etc. This makes sure that `remoteStdin` gets closed/reset as soon as the copy unblocks. Make sense?

liggitt replied:
interesting, the `cp` essentially "intercepts" the EOF, so the remote command never sees it.

ncdc replied:
Hmm, assuming I merged and rebuilt everything locally correctly, trying to exit an interactive shell is hanging. Grrr....

liggitt replied:
the deferred remoteStdin.Close() in the goroutine makes sense, but I would keep the deferred `Reset()` in place

ncdc replied:
Ok, added it back.

Commit:
Various exec fixes

If stdin is noninteractive, the io.Copy from stdin to remoteStdin will
unblock when it finishes reading from stdin. In this case, make sure to
close remoteStdin so the server knows the client won't be sending any
more data. This ensures that the remote process terminates. For example:

echo foo | kubectl exec -i <pod> -- cat

Without this change, the `cat` process never terminates and `kubectl
exec` hangs.

Fix interactive exec sessions hanging after you type 'exit'.

Add e2e test to cover noninteractive stdin: `echo a | kubectl exec -i <pod>
cat`

Add e2e test to cover psuedo-interactive stdin: `kubectl exec -i <pod> bash`

Prep for sending multiple data frames over multiple streams in remote command
test, which is more likely to find flakes (requires bump of spdystream
once an issue with the frame worker queues not being fully drained when
a goaway frame is received).

ncdc commented:
@liggitt @smarterclayton PR updated. I *think* I've finally gotten there...

liggitt commented:
LGTM... the additional e2e tests give me a lot more confidence

ncdc commented:
And e2e passed, woot!

ncdc commented:
... oh, for a previous commit. Guess we'll wait a bit longer

ncdc commented:
e2e failure looks like flakes:

```
[Fail] Services [It] should serve multiport endpoints from pods 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1147

[Fail] Services [It] should serve a basic endpoint from pods 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1147

[Fail] Services [It] should be able to up and down services 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:247
```

@k8s-bot test this please

liggitt commented:
wondering how to kick what appears to be a hung e2e... @k8s-bot test this again please?

ncdc commented:
Woot :-)
